sstables/sstable_set.hh

```diff
@@ -47,24 +47,86 @@ public:
 // Structure holds all sstables (a.k.a. fragments) that belong to same run identifier, which is an UUID.
 // SStables in that same run will not overlap with one another.
 class sstable_run {
-    sstable_list _all;
+    std::unordered_set<shared_sstable> _all;
 public:
+    sstable_run(std::unordered_set<shared_sstable> all) : _all(std::move(all)) {}
     void insert(shared_sstable sst);
     void erase(shared_sstable sst);
     // Data size of the whole run, meaning it's a sum of the data size of all its fragments.
     uint64_t data_size() const;
-    const sstable_list& all() const { return _all; }
+    const std::unordered_set<shared_sstable>& all() const { return _all; }
 };
 
-class sstable_set {
+class versioned_sstable_set_data {
```

Comment:
I think you can call it that. All members of this class contain all sstables from all versions that are currently alive

Reply:
I thought it only includes the sstables that all versions include, later additions being only found in those later versions.

Doesn't this mean that if we have two versions, each adding 1 sstable, then they both contain N-1 and N-2 sstables respectively?

Reply:
This versioned_sstable_set_data is shared across all versions, so every addition is still kept in its respective version. The shared data is mostly needed for two things: 
- the iterator given by sstable_list - we want to iterate over a single container because if we were iterating over containers from different versions, some versions could be merged and that would invalidate our iterators
- incremental selectors - currently each incremental_selector has references to members of an sstable_set, and we had different sstable_subsets for all versions, the incremental_selector would have to have references to all of them, and they could be merged too. These selectors also remember the last selected position, which would change during merging as well